Add function to write RunHeader.

JeremyMcCormick opened the pull request:
- Adds reading/writing of multiple RunHeader objects using EventFile.

- Adds lookup map of run number to RunHeader in EventFile.

- Adds destructor to EventFile.

- Adds Print() impl to RunHeader.

- Modifies event processing in Process class to read RunHeader objects from the EventFile (if files do not have the run header tree then there is an error which is trapped and a warning message is printed).

- Changes Process::onNewRun() method to accept a RunHeader reference instead of just the run number.

jmmans commented:
Questions on the Run Header overall:
  (1) Is this a _run_ header or a _file_ header?  
  (2) If it a run header, are events from multiple runs allowed in the same file?
  (3) If they are, we have to understand how to properly get the right run header for a given event (probably not too difficult, I suppose, by reading all the run headers when a file is opened, building a map and providing the right one for an event based on its run number.)
  (4) If it is a run header, we need to make sure we don't put in multiple copies when creating or copying files.

In terms of implementation, I would add a virtual method to Event which accesses the RunHeader.

JeremyMcCormick commented:
Thanks for the feedback, Jeremy.

Here are my assumptions:

- The RunHeader is a run header, not a file header.  It is a data object that is accessible through EventFile similarly to how events are managed.  There can be multiple RunHeader objects per file.

- There should/would be one RunHeader in the branch for every unique run number in the file.  The RunHeader would be duplicated across all files in the run so that the files are standalone.

- Rather than being accessed through the Event, I would imagine that the onNewRun() method of EventProcessor should accept a RunHeader parameter with the run info (detector name, parameters, etc.) instead of just the run number, as it does currently.

My plan is that EventFile would create a map of run number to RunHeader objects in its constructors.  Then there would be a function for getting a RunHeader by run number.

Process would feed new RunHeader objects to processors automatically as new run numbers are encountered in the job.

Let me know what you think of this plan.

JeremyMcCormick commented:
I've added functionality to EventFile to read/write multiple RunHeader objects from the run header branch.  I also changed the Process class so that the onNewRun() function takes a reference to a RunHeader object instead of just the run number (I think this will be more useful than just the run number).

Let me know what you think of the this implementation.

jmmans commented:
I like the implementation.  My only comment is that the overall strategy 
has been to avoid #including unnecessary header files -- can we make a 
forward declaration for RunHeader instead of #including it, when many 
users won't care/ever use the run header?

Jeremy

On 01/25/2017 05:20 PM, Jeremy McCormick wrote:
> I've added functionality to EventFile to read/write multiple RunHeader
> objects from the run header branch. I also changed the Process class so
> that the onNewRun() function takes a reference to a RunHeader object
> instead of just the run number (I think this will be more useful than
> just the run number).
>
> Let me know what you think of the this implementation.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/LDMXAnalysis/ldmx-sw/pull/144#issuecomment-275264045>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEtH4NbXY9hUmeccYsX0tqJt1Kb9l7yiks5rV9ijgaJpZM4LtB4s>.
>


-- 

Jeremiah (Jeremy) Mans
  Associate Professor of Physics     |  Email : jmmans@physics.umn.edu
  University of Minnesota            |          Jeremy.Mans@cern.ch
  PaN 427  / 612-625-8994  | "A witty saying proves nothing." - Voltaire
========================================================================

JeremyMcCormick commented:
I think a forward declaration can be used in EventFile.h because it uses a pointer to the RunHeader but the header needs to be included in EventProcessor.h as it is defined as a const ref.  I will go ahead and change this.

jmmans commented:
Hi Jeremy,

If it's a ref, it can be forward declared.

Jeremy

On 01/26/2017 01:26 PM, Jeremy McCormick wrote:
> I think a forward declaration can be used in EventFile.h because it uses
> a pointer to the RunHeader but the header needs to be included in
> EventProcessor.h as it is defined as a const ref. I will go ahead and
> change this.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/LDMXAnalysis/ldmx-sw/pull/144#issuecomment-275486647>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEtH4Dlehtb9r5_zAjuJA-GTiSFw8pTHks5rWPNmgaJpZM4LtB4s>.
>


-- 

Jeremiah (Jeremy) Mans
  Associate Professor of Physics     |  Email : jmmans@physics.umn.edu
  University of Minnesota            |          Jeremy.Mans@cern.ch
  PaN 427  / 612-625-8994  | "A witty saying proves nothing." - Voltaire
========================================================================

JeremyMcCormick commented:
I changed EventFile and EventProcessor to use forward declarations of RunHeader.

I also added a missing destructor to EventFile, where the RunHeader objects in the map are deleted (they are copied using "new" when being read from the input tree).

Are there any other objects that should be deleted or cleaned up in the destructor?
